$NetBSD: patch-SConstruct,v 1.3 2026/02/19 08:38:08 wiz Exp $

Add support for NetBSD and Dragonfly.
Fix locations.
Don't compile with debug info.
Don't mess with the linker.
Respect LDFLAGS and CXXFLAGS.
Fix compatibility issue with scons 4.9's CheckLibWithHeader
Fix build with setuptools 82 (pkg_resources removal).

--- SConstruct.orig	2025-12-22 23:06:05.000000000 +0000
+++ SConstruct
@@ -16,8 +16,6 @@ from glob import glob
 import uuid
 from glob import glob
 
-from pkg_resources import parse_version
-
 import SCons
 
 # This must be first, even before EnsureSConsVersion, if
@@ -1189,6 +1187,7 @@ envDict = dict(BUILD_ROOT=buildDir,
                CONFIGURELOG='$BUILD_ROOT/scons/config.log',
                CONFIG_HEADER_DEFINES={},
                LIBDEPS_TAG_EXPANSIONS=[],
+               ENV=os.environ,
                )
 
 # TODO: Remove these when hygienic builds are default.
@@ -1360,7 +1359,9 @@ os_macros = {
 os_macros = {
     "windows": "defined(_WIN32)",
     "solaris": "defined(__sun)",
+    "dragonfly": "defined(__DragonFly__)",
     "freebsd": "defined(__FreeBSD__)",
+    "netbsd": "defined(__NetBSD__)",
     "openbsd": "defined(__OpenBSD__)",
     "iOS": "defined(__APPLE__) && TARGET_OS_IOS && !TARGET_OS_SIMULATOR",
     "iOS-sim": "defined(__APPLE__) && TARGET_OS_IOS && TARGET_OS_SIMULATOR",
@@ -1852,7 +1853,7 @@ if env['_LIBDEPS'] == '$_LIBDEPS_LIBS':
     if not env.TargetOSIs('solaris', 'darwin', 'windows', 'openbsd'):
         env.Tool('thin_archive')
 
-if env.TargetOSIs('linux', 'freebsd', 'openbsd'):
+if env.TargetOSIs('linux', 'freebsd', 'openbsd', 'netbsd', 'dragonfly'):
     # NOTE: The leading and trailing spaces here are important. Do not remove them.
     env['LINK_WHOLE_ARCHIVE_LIB_START'] = '-Wl,--whole-archive '
     env['LINK_WHOLE_ARCHIVE_LIB_END'] = ' -Wl,--no-whole-archive'
@@ -1892,14 +1893,14 @@ elif env.TargetOSIs('solaris'):
 elif env.TargetOSIs('solaris'):
      env.Append( LIBS=["socket","resolv","lgrp"] )
 
-elif env.TargetOSIs('freebsd'):
+elif env.TargetOSIs('freebsd', 'dragonfly'):
     env.Append( LIBS=[ "kvm" ] )
     env.Append( CCFLAGS=[ "-fno-omit-frame-pointer" ] )
 
 elif env.TargetOSIs('darwin'):
     env.Append( LIBS=["resolv"] )
 
-elif env.TargetOSIs('openbsd'):
+elif env.TargetOSIs('netbsd', 'openbsd'):
     env.Append( LIBS=[ "kvm" ] )
 
 elif env.TargetOSIs('windows'):
@@ -2197,7 +2198,6 @@ if env.TargetOSIs('posix'):
     env.Append( CCFLAGS=["-fno-omit-frame-pointer",
                          "-fno-strict-aliasing",
                          "-fasynchronous-unwind-tables",
-                         "-ggdb" if not env.TargetOSIs('emscripten') else "-g",
                          "-pthread",
                          "-Wall",
                          "-Wsign-compare",
@@ -2860,9 +2860,9 @@ def doConfigure(myenv):
     if usingLibStdCxx:
         def CheckModernLibStdCxx(context):
             test_body = """
-            #if !__has_include(<experimental/filesystem>)
-            #error "libstdc++ from GCC 5.3.0 or newer is required"
-            #endif
+            //#if !__has_include(<experimental/filesystem>)
+            //#error "libstdc++ from GCC 5.3.0 or newer is required"
+            //#endif
             """
 
             context.Message('Checking for libstdc++ 5.3.0 or better... ')
@@ -3217,7 +3217,7 @@ def doConfigure(myenv):
         #
         myenv.Append( CCFLAGS=["/Zc:inline"])
 
-    if myenv.ToolchainIs('gcc', 'clang'):
+    if myenv.ToolchainIs('gcc', 'clang') and get_option('runtime-hardening') == "on":
         # This tells clang/gcc to use the gold linker if it is available - we prefer the gold linker
         # because it is much faster. Don't use it if the user has already configured another linker
         # selection manually.
@@ -3449,8 +3449,9 @@ def doConfigure(myenv):
     def checkOpenSSL(conf):
         sslLibName = "ssl"
         cryptoLibName = "crypto"
-        sslLinkDependencies = ["crypto", "dl"]
-        if conf.env.TargetOSIs('freebsd'):
+        if conf.env.TargetOSIs('linux', 'solaris'):
+            sslLinkDependencies = ["crypto", "dl"]
+        else:
             sslLinkDependencies = ["crypto"]
 
         if conf.env.TargetOSIs('windows'):
@@ -3499,7 +3500,7 @@ def doConfigure(myenv):
                 cryptoLibName,
                 ["openssl/crypto.h"],
                 "C",
-                "SSLeay_version(0);",
+                call="SSLeay_version(0);",
                 autoadd=True):
             maybeIssueDarwinSSLAdvice(conf.env)
             conf.env.ConfError("Couldn't find OpenSSL crypto.h header and library")
@@ -3656,7 +3657,7 @@ def doConfigure(myenv):
         if conf.CheckLibWithHeader(
             "curl",
             ["curl/curl.h"], "C",
-            "curl_global_init(0);",
+            call="curl_global_init(0);",
             autoadd=False):
             return True
 
@@ -3782,12 +3783,12 @@ def doConfigure(myenv):
             "sasl2",
             ["stddef.h","sasl/sasl.h"],
             "C",
-            "sasl_version_info(0, 0, 0, 0, 0, 0);",
+            call="sasl_version_info(0, 0, 0, 0, 0, 0);",
             autoadd=False ):
         myenv.ConfError("Couldn't find SASL header/libraries")
 
     # requires ports devel/libexecinfo to be installed
-    if env.TargetOSIs('freebsd', 'openbsd'):
+    if env.TargetOSIs('dragonfly', 'freebsd', 'netbsd', 'openbsd'):
         if not conf.CheckLib("execinfo"):
             myenv.ConfError("Cannot find libexecinfo, please install devel/libexecinfo.")
 
@@ -3947,7 +3948,7 @@ def doConfigure(myenv):
                 ["mongoc-1.0"],
                 ["mongoc/mongoc.h"],
                 "C",
-                "mongoc_get_major_version();",
+                call="mongoc_get_major_version();",
                 autoadd=False ):
             conf.env['MONGO_HAVE_LIBMONGOC'] = "library"
         if not conf.env['MONGO_HAVE_LIBMONGOC'] and env.TargetOSIs('darwin') and conf.CheckMongoCFramework():
@@ -4130,10 +4131,6 @@ if get_option('ninja') != 'disabled':
         env.SetOption('num_jobs', cpu_count)
 
 if get_option('ninja') != 'disabled':
-
-    if 'ICECREAM_VERSION' in env and not env.get('CCACHE', None):
-        if env['ICECREAM_VERSION'] < parse_version("1.2"):
-            env.FatalError("Use of ccache is mandatory with --ninja and icecream older than 1.2. You are running {}.".format(env['ICECREAM_VERSION']))
 
     ninja_builder = Tool("ninja")
     env["NINJA_BUILDDIR"] = env.Dir("$BUILD_DIR/ninja")
